lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1313,11 +1323,26 @@ static void diagSyntacticUseRestrictions(const Expr *E, const DeclContext *DC,
 
   // Diagnose uses of collection literals with defaulted types at the top
   // level.
-  if (auto collection
-        = dyn_cast<CollectionExpr>(E->getSemanticsProvidingExpr())) {
+  E = E->getSemanticsProvidingExpr();
+  bool diagnoseAsWarning = false;
+
+  // Look through implicit conversions.
+  if (auto conversion = dyn_cast<ImplicitConversionExpr>(E)) {
+    // This case was not diagnosed previously; emit a warning instead of an
+    // error for source compatibility.
+    diagnoseAsWarning = true;
+    while (auto *expr =
+               dyn_cast<ImplicitConversionExpr>(conversion->getSubExpr())) {
+      conversion = expr;
+    }
+
+    E = conversion->getSubExpr();
+  }
+
+  if (auto collection = dyn_cast<CollectionExpr>(E)) {
     if (collection->isTypeDefaulted()) {
-      Walker.checkTypeDefaultedCollectionExpr(
-        const_cast<CollectionExpr *>(collection));
+      Walker.diagnoseTypeDefaultedCollectionExpr(
```

Comment:
I think we should go one step farther with this and lift the check to somewhere around https://github.com/apple/swift/blob/main/lib/Sema/MiscDiagnostics.cpp#L133 which would mean that inner defaulted collections are going to be diagnosed as well ad we don't have to look through `ImplicitConversionExpr` explicitly since that would already be handled by the walker.

Here are a couple of examples of what needs to warn but doesn't at the moment:

```
func test<T>(_: [T]) {}

test([]) // should warn because T is defaulted to Any 
test([1, nil, ""]) // same issue
```

and

```
func test<T>() -> [T]? {
  []
}

_ = test() ?? [] // same here T is defaulted to Any from literal collection
```

Reply:
@xedin 
Would this be another example as well?
```swift
func f(_ b: Bool) {
  let _ = b ? [] : []
}
```

Reply:
Yes!

Reply:
Thank you for the review! Once I moved the functionality to the upper-level, I ran into a problem where inner arrays in expressions like `let a: [Any]  = [1, nil, [2, "Swift"]]` are diagnosed, even though they shouldn't. I'm working on restricting emission of any diagnostics in this case, since I can't think of any better solution.

Reply:
That happens because we're walking into the first array and then we see that nested array, which is inferred as `[Any]` and diagnostics gets emitted.

Reply:
That is actually correct, the second array is heterogeneous as well.

Reply:
Okay, but I thought we can't emit diagnostics for nested stuff in arrays (at least my mentor told me so)

Reply:
I don’t see why not in this case but you could just check whether `Parent` is a collection expression and if so - skip this diagnostic.

Reply:
Maybe only ASTVisitor tracks Parent so you might need to write some code it track it in here.

Reply:
Yea, this makes sense!

Reply:
Ok, I was got it right - `ASTWalker` has `Parent` it should be easy. If you are curious about the direct tracking good example of such pattern could be found in PreCheckExpr which needs to track current declaration context because it walks into closures. 

So before walking into the closure body it would switch `DC` to a closure - https://github.com/apple/swift/blob/main/lib/Sema/PreCheckExpr.cpp#L1469 

and then when walking out back to its parent -  https://github.com/apple/swift/blob/main/lib/Sema/PreCheckExpr.cpp#L1310-L1313

Reply:
> I thought we can't emit diagnostics for nested stuff in arrays (at least my mentor told me so)

For context, this was me trying to make sense of [this](https://github.com/apple/swift/blob/6c7e8c4c5bdc3c0086a4f149b39efe526a8744c5/lib/Sema/MiscDiagnostics.cpp#L425) comment. It seems like we do not want to diagnose a nested collection literal if we already diagnosed the outer literal, but I am not certain.

Reply:
IMHO it just states what implementation does which doesn’t account for the fact that the types flow both ways and erasure/defaulting might happen in the inner collection without outer being defaulted or erased…

Reply:
@xedin I just pushed new version of the changes. The check had been lifted to Walker

Reply:
I don't see why we shouldn't diagnose nested literals either, even when both the outer and inner literals are defaulted.